### PR TITLE
Story 4 game animation

### PR DESCRIPTION
**Pull request by f-gate**
to be reviewed:
anims-missiles.js -- run the function inside the console as the event handler does not exist to start game.
e.g: dropMissile(2, 1000)

### CODE REVIEW

**Comment by ucanfil** [comment]
Looks good.

**Review by ucanfil — DISMISSED** [review]
Looks good.

File: JS/anims-missiles.js [hunk 1]
```diff
@@ -1,11 +1,25 @@
...
     }, animationTime, "linear", function () {
         $(this).remove()
-
     })
 }
-// run on start button
\ No newline at end of file
+
+function randomNumGen() {
+    var randomNum = Math.floor(Math.random() * 4) + 1
```
Comment (patrickkenyon) [inline JS/anims-missiles.js]: Could use the below:
`Math.ceil(Math.random() * 4)`

File: JS/anims-missiles.js [hunk 2]
```diff
@@ -1,11 +1,25 @@
-function dropMissile(columnNumber, animationTime) {
+setTimeout(repeatAnims, 500)
+repeatAnims()
+
+function repeatAnims() {
+    setInterval( function () {
+        dropMissile(1000)
+    }, 1000 )
```
Comment (patrickkenyon) [inline JS/anims-missiles.js]: whitespace

File: JS/anims-missiles.js [hunk 3]
```diff
@@ -1,11 +1,25 @@
...
+}
+
+function dropMissile(animationTime) {
+    var columnNumber = randomNumGen()
     $('#pos-'+columnNumber).prepend('<img src="img/missile-drop-down-white.png" class="missile" id="bomb-'+ columnNumber +'">')
```
Comment (patrickkenyon) [inline JS/anims-missiles.js]: Line over 80 characters add break and two tabs for next line

**Review by patrickkenyon — COMMENTED** [review]
Just a few comments

File: JS/anims-missiles.js (hunk 1 above)
Comment (imogenhallett) [reply JS/anims-missiles.js]: This has been fixed in 6181f9b

File: JS/anims-missiles.js (hunk 2 above)
Comment (imogenhallett) [reply JS/anims-missiles.js]: fixed in 973e72b

**Review by imogenhallett — COMMENTED** [review]
Resolved in 45b1c17

File: JS/anims-missiles.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    $('#bomb-' + columnNumber).animate({
+        top: "+=550"
+    }, animationTime, "linear", function () {
+        console.log($(this))
+        $(this).remove()
+        document.querySelector('#toll').innerHTML = incrementToll()
+    })
+}
+
+function randomNumGen() {
```
Comment (mporam) [inline JS/anims-missiles.js]: not a great function name. it doesnt generate a random number, it generates a random number between 1 and 6

File: JS/anims-missiles.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+function dropMissile(animationTime) {
+    var columnNumber = randomNumGen()
+    $('#pos-'+columnNumber).prepend('<img src="img/missile-drop-down-white.png" ' +
+        'class="missile" id="bomb-'+ columnNumber +'">')
+
+    $('#bomb-' + columnNumber).animate({
+        top: "+=550"
+    }, animationTime, "linear", function () {
+        console.log($(this))
```
Comment (mporam) [inline JS/anims-missiles.js]: remove this

File: JS/audio.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+            var $city = $('#city-' + keys[e.key])
+            shootLaser($city)
+        }
+    })
+}
+
+var cities = document.querySelectorAll('.city')
+cities.forEach(function(city) {
+    city.addEventListener('click', function() {
+        var param = 'cheer' + city.dataset.city
```
Comment (mporam) [inline JS/audio.js]: param means nothing. you fixed the one above, why not this one??

File: js/anims-lasers.js [hunk 7]
```diff
@@ -0,0 +1,14 @@
+// $(function() {
+//     $('.city').click(function () {
+//         shootLaser($(this))
+//     })
+// })
```
Comment (mporam) [inline js/anims-lasers.js]: please delete rather than comment

**Review by mporam — CHANGES_REQUESTED** [review]

File: JS/anims-missiles.js (hunk 4 above)
Comment (imogenhallett) [reply JS/anims-missiles.js]: We are now using this function more generically to generate various random numbers up 10,000 so we would like to keep the function name the same.

**Comment by imogenhallett** [comment]
Please review

anims-missiles.js

File: JS/audio.js (hunk 6 above)
Comment (imogenhallett) [reply JS/audio.js]: this has been fixed

File: JS/anims-missiles.js (hunk 5 above)
Comment (imogenhallett) [reply JS/anims-missiles.js]: done

Comment (imogenhallett) [reply JS/anims-missiles.js]: done

File: js/anims-missiles.js [hunk 8]
```diff
@@ -0,0 +1,111 @@
+//global scope
+
+var columnID = 0
+
+
```
Comment (Benson-J) [inline js/anims-missiles.js]: Unnecessary whitespace

File: js/anims-missiles.js [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+function repeatAnims() {
+    setInterval(function() {
+        dropMissile(2000)
+    }, 1000)
+}
+
+function dropMissile(animationTime) {
+    var columnNumber = randomNumGen(4)
+    var $bomb = $('<img src="img/missile-drop-down-white.png" ' +
+        'class="missile bomb-'+ columnNumber +'">')
```
Comment (Benson-J) [inline js/anims-missiles.js]: Missile *AND* bomb!?

File: js/anims-missiles.js [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    $('#pos-'+columnNumber).prepend($bomb)
+
+    $bomb.animate({
+        top: "+=550"
+    }, animationTime, "linear", function () {
+        missileHitsCity()
+        $(this).remove()
+        document.querySelector('#toll').innerHTML = incrementToll()
```
Comment (Benson-J) [inline js/anims-missiles.js]: No longer necessary as death toll is handled in missileHitsCity(), which is called earlier in this function.

File: js/anims-missiles.js [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    })
+}
+
+function missileHitsCity() {
+    gameScore.score -= 1
+    gameScore.toll += randomNumGen(10000)
```
Comment (Benson-J) [inline js/anims-missiles.js]: Random number means you can't use it as a lives system, if we were to implement that in future.

File: js/anims-missiles.js [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+function isHit() {
+    var cityNumber = columnID
+    var bombNumber = '.bomb-' + cityNumber
+    var hasBombs = document.querySelectorAll(bombNumber).length
+    if (hasBombs) {
+        gameScore.score += hasBombs
+        document.querySelector('#score').textContent = gameScore.score
+        $(bombNumber).stop()
+        var bombs = document.querySelectorAll(bombNumber)
```
Comment (Benson-J) [inline js/anims-missiles.js]: Better to declare functions at top of scope.

File: js/anims-missiles.js [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+        }
+    })
+}
+
+createBombEventHandler()
+
+listenKeypressMissiles()
+
+
+function incrementToll() {
```
Comment (Benson-J) [inline js/anims-missiles.js]: No longer necessary, toll is handled elsewhere.

File: js/anims-missiles.js (hunk 8 above)
Comment (ucanfil) [reply js/anims-missiles.js]: Solved

File: js/anims-missiles.js (hunk 9 above)
Comment (ucanfil) [reply js/anims-missiles.js]: All missiles now

File: js/anims-missiles.js (hunk 10 above)
Comment (ucanfil) [reply js/anims-missiles.js]: removed

File: js/anims-missiles.js (hunk 11 above)
Comment (ucanfil) [reply js/anims-missiles.js]: have made it a constant 10,000 per miss.

File: js/anims-missiles.js (hunk 12 above)
Comment (ucanfil) [reply js/anims-missiles.js]: done

File: js/anims-missiles.js (hunk 13 above)
Comment (ucanfil) [reply js/anims-missiles.js]: this has been removed

File: js/anims-missiles.js [hunk 14]
```diff
@@ -58,21 +54,21 @@ function createBombEventHandler() {
 
 function isHit() {
     var cityNumber = columnID
```
Comment (Benson-J) [inline js/anims-missiles.js]: Redundant variable.

**Review by Benson-J — COMMENTED** [review]
All good.

**Comment by ucanfil** [comment]
Hi Mike, 
please review all of  js/anims-missiles.js

we are aware we need to add comments but wanted to get the bulk of the code to you ASAP.

File: js/anims-missiles.js [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+            isHit()
+        })
+    })//end forEach
+
+}
+
+function isHit() {
+    var missileNumber = '.missile-' + columnID
+    var missiles = document.querySelectorAll(missileNumber)
+    var hasMissiles = document.querySelectorAll(missileNumber).length
```
Comment (mporam) [inline js/anims-missiles.js]: this is the same selector twice. why not do:
` var hasMilssiles = missiles.length`

File: js/anims-missiles.js [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+    var columnNumber = randomNumGen(4)
+    var $missile = $('<img src="img/missile-drop-down-white.png" ' +
+        'class="missile missile-'+ columnNumber +'">')
+    $('#pos-'+columnNumber).prepend($missile)
+
+    $missile.animate({
+        top: "+=550"
+    }, animationTime, "linear", function () {
+        missileHitsCity()
+        $(this).remove()
```
Comment (mporam) [inline js/anims-missiles.js]: why not put this inside missileHitsCity()? you will always want to remove the missile when it hits

File: js/anims-missiles.js [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+function missileHitsCity() {
+    gameScore.score -= 1
+    gameScore.toll += 10000
+    document.querySelector('#score').textContent = gameScore.score
+    document.querySelector('#toll').textContent = gameScore.toll
+}
+
+function randomNumGen(topLimit) {
+    var randomNum = Math.ceil(Math.random() * topLimit)
+    return randomNum
```
Comment (mporam) [inline js/anims-missiles.js]: just return the line above, no need to create the variable, wasting JS memory

File: js/anims-missiles.js [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+            'r': '4'
+        }
+        if (e.key in keys) {
+            columnID = keys[e.key]
+            isHit()
+        }
+    })
+}
+
+createMissileEventHandler()
```
Comment (mporam) [inline js/anims-missiles.js]: shouldn't these happen on start button key press?

File: js/anims-missiles.js [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+
+function isHit() {
+    var missileNumber = '.missile-' + columnID
+    var missiles = document.querySelectorAll(missileNumber)
+    var hasMissiles = document.querySelectorAll(missileNumber).length
+    if (hasMissiles) {
+        gameScore.score += hasMissiles
+        document.querySelector('#score').textContent = gameScore.score
+        $(missileNumber).stop()
+        missiles.forEach(function (missile) {
```
Comment (mporam) [inline js/anims-missiles.js]: consistency in your anonymous functions. i prefer `function(param)` but `function (param)` is fine, just make it consistent

**Review by mporam — CHANGES_REQUESTED** [review]

File: js/anims-missiles.js (hunk 15 above)
Comment (ucanfil) [reply js/anims-missiles.js]: Fixed in fa7a008

File: js/anims-missiles.js (hunk 16 above)
Comment (ucanfil) [reply js/anims-missiles.js]: Fixed in 80513ff

Had to pass the $(this) value into missileHitsCity() as 'that' because 'this' is a reserved word and therefore cannot pass 'this' as a parameter.

File: js/anims-missiles.js (hunk 17 above)
Comment (ucanfil) [reply js/anims-missiles.js]: Fixed in fc27f2a

File: js/anims-missiles.js (hunk 18 above)
Comment (ucanfil) [reply js/anims-missiles.js]: Yes, we are going to bundle all of our functions into one init() but is being called at the moment for testing purposes.

File: js/anims-missiles.js (hunk 19 above)
Comment (ucanfil) [reply js/anims-missiles.js]: Fixed in c4ad8f9

File: js/anims-missiles.js [hunk 20]
```diff
@@ -0,0 +1,116 @@
+//global scope
+var columnID = 0
+var gameScore = {
+    score:0,
+    toll:0
+}
+
+/**
+ * Function to prepend a misssle to the html and then animate top to bottom
```
Comment (charliecog) [inline js/anims-missiles.js]: missle spello

File: js/anims-missiles.js [hunk 21]
```diff
@@ -0,0 +1,116 @@
+//global scope
```
Comment (charliecog) [inline js/anims-missiles.js]: probably dont need to identify this

File: js/anims-missiles.js [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+var gameScore = {
+    score:0,
+    toll:0
+}
+
+/**
+ * Function to prepend a misssle to the html and then animate top to bottom
+ * @param int animationTime the speed at which to drop a single missle
+ */
+
```
Comment (charliecog) [inline js/anims-missiles.js]: remove line break

File: index.html [hunk 23]
```diff
@@ -71,7 +67,8 @@ <h2>DEATH TOLL:</h2>
         </div>
     </main>
```
Comment (charliecog) [inline index.html:95]: has this been tested in ie? ```<main>``` usually has problems

**Review by charliecog — COMMENTED** [review]
few little things to tweak/look at, looks good though!

File: js/anims-missiles.js (hunk 20 above)
Comment (ucanfil) [reply js/anims-missiles.js]: Fixed in f312209

File: js/anims-missiles.js (hunk 21 above)
Comment (ucanfil) [reply js/anims-missiles.js]: fixed in 2415fa8

File: js/anims-missiles.js (hunk 22 above)
Comment (ucanfil) [reply js/anims-missiles.js]: fixed in d5b10eb

**Comment by ucanfil** [comment]
Please review animationChangeSpeed function at line 119, and implementation of it in line 30/31
Already peer reviewed, please ignore pull request 'Story 4 speedup anims' as we're now working in this branch.

File: js/anims-missiles.js [hunk 24]
```diff
@@ -13,54 +16,71 @@ function dropMissile(animationTime) {
...
+        dropMissile(animationTime)
     }, 500)
 }
 
-function missileHitsCity() {
+/**
+ * Function decrements game score and increments death toll and renders the scores on the page when a city is hit
+ * @param that Had to pass the $(this) value into missileHitsCity() as 'that' because 'this' is a reserved word and therefore cannot pass 'this' as a parameter
+ */
+function missileHitsCity(that) {
```
Comment (mporam) [inline js/anims-missiles.js]: that isnt a great variable name. its the missile, why not use $missile

File: js/anims-missiles.js [hunk 25]
```diff
@@ -89,6 +112,20 @@ function listenKeypressMissiles() {
...
+/**
+ * Increases animation speed based upon gameScore.score
+ * We call this function inside repeatAnim function
+ */
+function animationChangeSpeed() {
+    var animationDuration = 3000
+    if (gameScore.score > 0) {
+        animationDuration = 3000 / (1 + (gameScore.score * 0.03))
+    } else {
+        animationDuration = 3000
```
Comment (mporam) [inline js/anims-missiles.js]: this else does nothing

File: js/anims-missiles.js [hunk 26]
```diff
@@ -89,6 +112,20 @@ function listenKeypressMissiles() {
...
 
-createMissileEventHandler()
+/**
+ * Increases animation speed based upon gameScore.score
+ * We call this function inside repeatAnim function
+ */
+function animationChangeSpeed() {
+    var animationDuration = 3000
+    if (gameScore.score > 0) {
+        animationDuration = 3000 / (1 + (gameScore.score * 0.03))
```
Comment (mporam) [inline js/anims-missiles.js]: why not just return this line instead of putting it in a variable

File: js/anims-missiles.js [hunk 27]
```diff
@@ -89,6 +112,20 @@ function listenKeypressMissiles() {
...
+    } else {
+        animationDuration = 3000
+    }
+    return animationDuration
```
Comment (mporam) [inline js/anims-missiles.js]: if you return line 122, you can return 3000 here

**Review by mporam — CHANGES_REQUESTED** [review]

File: js/anims-missiles.js (hunk 24 above)
Comment (ucanfil) [reply js/anims-missiles.js]: Fixed in 2cb838a

File: js/anims-missiles.js (hunk 25 above)
Comment (ucanfil) [reply js/anims-missiles.js]: Fixed in 193f632

File: js/anims-missiles.js (hunk 26 above)
Comment (ucanfil) [reply js/anims-missiles.js]: Fixed in 41ec220

File: js/anims-missiles.js (hunk 27 above)
Comment (ucanfil) [reply js/anims-missiles.js]: Switched with ternary operator in 0bfeea3

File: js/audio.js [hunk 28]
```diff
@@ -23,22 +23,26 @@ function shootLaser($city) {
     }, 300)
 }
 function listenCityClick() {
-    document.querySelectorAll('.city').forEach(function(city) {
+    var cities = Array.prototype.slice.call(document.querySelectorAll('.city'))
```
Comment (mporam) [inline js/audio.js]: not consistent with your other foreachs

**Review by mporam — CHANGES_REQUESTED** [review]

**Comment by hoptond** [comment]
Completed the great smergening. Requesting review of the entire code at this point before it goes into master.

File: .gitignore [hunk 29]
```diff
@@ -1,2 +1,4 @@
 .idea/
 .DS_Store
+audio/hit_city_orig.wav
+img/city-hit-blank.gif
```
Comment (Benson-J) [inline .gitignore]: Do these need to be here?

File: css/game.css [hunk 30]
```diff
@@ -82,14 +82,37 @@ body {
...
+}
+
 #splash-background {
     position: absolute;
     width: 1000px;
 }
 
+#instructions-tab.active,
+#attribution-tab.active {
+    color: hotpink;
```
Comment (Benson-J) [inline css/game.css]: Use rgba or hex code for consistency.

File: css/game.css [hunk 31]
```diff
@@ -210,14 +246,22 @@ img.missile {
 }
 
 .start {
-    display: none;
+    position: absolute;
+    right: 5%;
+    bottom: 1%;
```
Comment (Benson-J) [inline css/game.css]: Use pixels not %.

File: .gitignore (hunk 29 above)
Comment (hoptond) [reply .gitignore]: removed in 534f7f9

File: css/game.css (hunk 31 above)
Comment (hoptond) [reply css/game.css]: fixed in e9ec040

File: css/game.css (hunk 30 above)
Comment (hoptond) [reply css/game.css]: fixed in 1ee20ff

**Review by Benson-J — DISMISSED** [review]

File: css/game.css [hunk 32]
```diff
@@ -135,6 +157,14 @@ img.missile {
     background-size: cover;
 }
 
+.explosion {
+    position: absolute;
+    bottom: -15px;
+    left: 0;
+    width: 100%;
+
```
Comment (mporam) [inline css/game.css]: whitespace

File: css/game.css [hunk 33]
```diff
@@ -214,14 +246,22 @@ img.missile {
...
     background-color: #b21123;
     color: #ffffff;
+    font-size: 2.4rem;
     margin: 0 auto;
     border-radius: 100%;
     border: 4px solid #cccccc;
     width: 100px;
     height: 100px;
     line-height: 90px;
     text-align: center;
```
Comment (mporam) [inline css/game.css:261]: can we have some padding on this button, the text looks a bit odd

File: js/anims-missiles.js [hunk 34]
```diff
@@ -59,41 +64,69 @@ function createMissileClickHandler() {
...
-    })//end forEach
+        city.addEventListener('click', checkForMissiles)
+    })
 }
+
+/**
+ * This was previously an anonymous function given its own function. It sets the column id and executes the isHit()
+ * method to scan for any missiles in the column
+ */
+function checkForMissiles() {
```
Comment (mporam) [inline js/anims-missiles.js]: this function doesnt check for missiles, not sure the function name is correct

File: js/anims-missiles.js [hunk 35]
```diff
@@ -108,18 +141,41 @@ function listenKeypressMissiles() {
...
+    impacts[cityid - 1] = setTimeout(function (e) {
+        explosion.setAttribute('src', 'img/blank.png')
+    }, 1250)
+}
+
+document.querySelector('#start').addEventListener('click', function(e) {
+    repeatAnims()
+    missilesActive = true
+    var cities = document.querySelector('.city')
+     Array.prototype.forEach.call(cities, function(city) {
```
Comment (mporam) [inline js/anims-missiles.js]: indentation is incorrect here

File: js/anims-missiles.js [hunk 36]
```diff
@@ -59,41 +64,69 @@ function createMissileClickHandler() {
...
+}
+
 /**
  * If a missile is hit stop the anim and replace the missile img with explosion img - increment the deflects score.  If it is a miss decrement the deflects score and render on the page
  */
 function isHit() {
     var missileNumber = '.missile-' + columnID
     var missiles = document.querySelectorAll(missileNumber)
     var hasMissiles = missiles.length
+    var success = true
```
Comment (mporam) [inline js/anims-missiles.js:89]: why have this variable? you can simplify your code by removing this and just calling makeNoise in the relevant if, then you dont need the `laserResultSound` function

File: js/anims-missiles.js [hunk 37]
```diff
@@ -108,18 +141,41 @@ function listenKeypressMissiles() {
...
 }
 
-/**
- * Increases animation speed based upon gameScore.score
- * We call this function inside repeatAnim function
+/*
+ * This changes the speed of the missiles based upon the player's current score.
  */
 function animationChangeSpeed() {
     return gameScore.score > 0 ? 3000 / (1 + (gameScore.score * 0.03)) : 3000
```
Comment (mporam) [inline js/anims-missiles.js:154]: technically thius means it will get infinitely faster, with no upper limit. i.e untill the missiles are falling every millisecond (1000 per second) is that what you want?

File: js/audio.js [hunk 38]
```diff
@@ -1,48 +1,18 @@
-function listenKeypress() {
```
Comment (mporam) [inline js/audio.js:1]: this file is now so small is it worth having it? a http request just to get two functions that are only used from anims-missiles.js - which should be renamed also as it does a lot more than animate the missiles now

File: js/audio.js [hunk 39]
```diff
@@ -1,48 +1,18 @@
...
-            var $city = $('#city-' + keys[e.key])
-            shootLaser($city)
-        }
-    })
-}
-
 function shootLaser($city) {
-    var cityid = $city.data('city')
     $city.addClass('laser')
     setTimeout(function () {
```
Comment (mporam) [inline js/audio.js]: `function()`

File: js/anims-missiles.js [hunk 40]
```diff
@@ -108,18 +141,41 @@ function listenKeypressMissiles() {
...
+ *
+ * @param cityid The city being exploded.
+ */
+function hitCityEffect(cityid) {
+    makeNoise('hitcity')
+    var explosion = document.querySelector('#city-' + cityid + ' .explosion')
+    clearTimeout(impacts[cityid - 1])
+    //we append a random query string to make the gif reliably start at zero when a new explosion occurs.
+    explosion.setAttribute('src', 'img/city-hit.gif' + '?explode=' + Math.random(4))
+    impacts[cityid - 1] = setTimeout(function (e) {
```
Comment (mporam) [inline js/anims-missiles.js]: `function(`

**Review by mporam — CHANGES_REQUESTED** [review]

File: css/game.css (hunk 32 above)
Comment (hoptond) [reply css/game.css]: fixed in bffb0eb

File: css/game.css (hunk 33 above)
Comment (hoptond) [reply css/game.css:261]: changed in a09d7d0

File: js/anims-missiles.js (hunk 34 above)
Comment (hoptond) [reply js/anims-missiles.js]: fixed in eb98601

File: js/anims-missiles.js (hunk 36 above)
Comment (hoptond) [reply js/anims-missiles.js:89]: this function is called both during the demo phase and during the actual game phase. We needed to overcome an earlier design desicion and this was the best way we could come up with in the time we had.

File: js/anims-missiles.js (hunk 37 above)
Comment (hoptond) [reply js/anims-missiles.js:154]: Yes, because the game will start to get so hard the user will miss which will result in the score decreasing and thus the speed will go back to something the user is capable of dealing with.

File: js/audio.js (hunk 38 above)
Comment (hoptond) [reply js/audio.js:1]: addressed in 038d9af

File: js/anims-missiles.js (hunk 40 above)
Comment (hoptond) [reply js/anims-missiles.js]: fixed in 2efc0aa

File: js/audio.js (hunk 39 above)
Comment (hoptond) [reply js/audio.js]: fixed in 2efc0aa

File: js/anims-missiles.js (hunk 35 above)
Comment (hoptond) [reply js/anims-missiles.js]: fixed in 4a1005c

**Review by mporam — APPROVED** [review]
